### PR TITLE
[lte][gateway][focal] add ubuntu focal fossa vagrant vm

### PR DESCRIPTION
**Pull request by kkahrs**
Signed-off-by: Ken Kahrs <kkahrs@gmail.com>

## Summary
Add experimental Ubuntu Focal (20.04) VM to lte/gateway

## Test Plan

* verify `vagrant up magma_focal` completes
* verify `vagrant up magma` completes
* verify `fab dev package:vcs=git` still succeeds on original `magma` VM

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Comment by 119Vik** [comment]
RestartCWAGTests

**Comment by 119Vik** [comment]
RestartCWAGTests

**Comment by 119Vik** [comment]
RestartCWAGTests

**Comment by 119Vik** [comment]
RestartCWAGTests

**Comment by 119Vik** [comment]
RestartCWAGTests

**Comment by 119Vik** [comment]
RestartCWAGTests

File: lte/gateway/deploy/magma_dev_focal.yml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      vars:
+        c_build: /home/{{ ansible_user }}/build/c/
+        oai_build: "{{ c_build }}/oai"
+        go_build: /home/{{ ansible_user }}/go/bin/
+        magma_repo: /home/{{ ansible_user }}/magma-packages
+        magma_deps: /home/{{ ansible_user }}/magma-deps
+    - role: fluent_bit
+    - role: envoy
+    - role: pyvenv
+    - role: focal_hacks
```
Comment (tmdzk) [inline lte/gateway/deploy/magma_dev_focal.yml:53]: where is the `focal_hacks` role? Maybe we should rename it to focal_specific something like that

File: lte/gateway/deploy/roles/magma/tasks/main.yml [hunk 2]
```diff
@@ -163,26 +164,51 @@
...
 
 - name: Overwrite linux 4.9 GTP module with patched module
   apt: pkg=oai-gtp=4.9-5 state=present update_cache=yes dpkg_options=force-overwrite
   retries: 5
-  when: preburn
+  when: preburn and ansible_distribution == "Debian"
 
-- name: Install Magma dependencies
+- name: Install common Magma dependencies
+  retries: 5
```
Comment (tmdzk) [inline lte/gateway/deploy/roles/magma/tasks/main.yml:201]: same here

File: lte/gateway/deploy/roles/magma/tasks/main.yml [hunk 3]
```diff
@@ -145,12 +145,13 @@
   when: full_provision
 
 - name: Install OpenAirInterface (OAI) dependencies
+  retries: 5
```
Comment (tmdzk) [inline lte/gateway/deploy/roles/magma/tasks/main.yml:148]: Do we really need that? seems like a hack. maybe file a task to fix this no?

File: orc8r/tools/ansible/roles/fluent_bit/tasks/main.yml [hunk 4]
```diff
@@ -18,7 +18,7 @@
 
 - name: Add fluent bit repo to /etc/apt/sources.list
   apt_repository:
-    repo: deb https://packages.fluentbit.io/debian/stretch stretch main
+    repo: deb https://packages.fluentbit.io/{{ ansible_distribution|lower}}/{{ ansible_distribution_release }} {{ ansible_distribution_release }} main
```
Comment (tmdzk) [inline orc8r/tools/ansible/roles/fluent_bit/tasks/main.yml:21]: I would leave it that way or just move it to the var file as a single Variable

File: third_party/build/patches/libfluid/libfluid_base_patches/EVLOOP_NO_EXIT_ON_EMPTY_compat.patch [hunk 5]
```diff
@@ -0,0 +1,36 @@
+diff --git a/fluid/base/EventLoop.cc b/fluid/base/EventLoop.cc
```
Comment (tmdzk) [inline third_party/build/patches/libfluid/libfluid_base_patches/EVLOOP_NO_EXIT_ON_EMPTY_compat.patch:1]: Where does that patch come from stackoverflow? Is that the only way to get it? (maybe cherry-pick if a commit has been created)

**Review by tmdzk — COMMENTED** [review]
just a few comments. Otherwise gj

File: lte/gateway/deploy/roles/magma/tasks/main.yml (hunk 3 above)
Comment (kkahrs) [reply lte/gateway/deploy/roles/magma/tasks/main.yml:148]: not really added -- was present in original task, which was copied as a template for the new per-distribution tasks

File: orc8r/tools/ansible/roles/fluent_bit/tasks/main.yml (hunk 4 above)
Comment (kkahrs) [reply orc8r/tools/ansible/roles/fluent_bit/tasks/main.yml:21]: this change uses ansible facts to deduce the correct repo url for fluentbit -- becomes
`deb https://packages.fluentbit.io/ubuntu/focal focal main`
on ubuntu focal, or
`deb https://packages.fluentbit.io/debian/stretch stretch main`
on debian stretch

should stay this way unless it causes a problem for e.g. Red Hat

File: lte/gateway/deploy/magma_dev_focal.yml (hunk 1 above)
Comment (kkahrs) [reply lte/gateway/deploy/magma_dev_focal.yml:53]: added after move from later commit... the name `focal_hacks` indicates tasks (only 1 so far) where a shortcut was taken to get the vm to provision and do something useful but should be fixed for real in the near future

File: third_party/build/patches/libfluid/libfluid_base_patches/EVLOOP_NO_EXIT_ON_EMPTY_compat.patch (hunk 5 above)
Comment (kkahrs) [reply third_party/build/patches/libfluid/libfluid_base_patches/EVLOOP_NO_EXIT_ON_EMPTY_compat.patch:1]: no, patch was my own work (originally for CentOS 8 experimental build) -- would be great if we could get libfluid to import it

**Comment by pshelar** [comment]
I could build magma packages with the PR on latest master, `fab dev package:git`.

**Review by tmdzk — APPROVED** [review]
💃

**Comment by pshelar** [comment]
This only required for debian stretch
```
- name: Preload the GTP module
  ignore_errors: yes
  modprobe:
    name: gtp
    state: present
  when: full_provision```

**Pull request merged**